src/Bitmap_cubical_complex/include/gudhi/embedded_complex_release.hpp

```diff
@@ -0,0 +1,484 @@
+
+#include <future>
+#include <thread>
+#include <chrono>
+
+#ifndef EMBEDDED_CPLX
+#define EMBEDDED_CPLX
+
+template <typename T>
+void print_vector(std::vector<T> vect){
+    if(vect.size() == 0){
+        std::cout << "[]\n";
+    }else{
+        std::cout << "[" << vect[0];
+        for(long unsigned int i=1; i<vect.size(); i++){
+            std::cout << ", " << vect[i];
+        }
+        std::cout << "]\n";
+    }
+}
+
+template <typename T>
+class Embedded_cubical_complex : public Gudhi::cubical_complex::Bitmap_cubical_complex<T>
+{
+
+    private:
+
+        typedef std::size_t Simplex_key;
+        typedef typename T::filtration_type Filtration_value;
+        typedef Simplex_key Simplex_handle;
+
+    protected:
+        std::vector<std::vector<double>> embedding;         //Array of the cubical complexe's points' coordinates
+        std::vector<int> embedding_index;                   //Array to link vertices index in the cubical complex to their index in the embedding
+
+        std::vector<int> sizes_pdt;                         //Products of the sizes from s_0 up to s_0*s_1*...*s_(n-2)
+
+        std::vector<std::vector<int>> critical_vertices;
+        std::vector<std::vector<int>> critical_multiplicity;
+
+        //If we have a vector e = (e_1,...,e_n), its index is the sum of the 2^i were i are the indexes such that e_i >= 0
+        //If index_v = index_w they have the same critical points. The critical points are stored in critical_vertices[index]
+
+    public:
+        //*********************************************//
+        //Constructors
+        //*********************************************//
+        Embedded_cubical_complex(const std::vector<unsigned>& dimensions,
+            const std::vector<Filtration_value>& top_dimensional_cells, unsigned int num_jobs = 0):Gudhi::cubical_complex::Bitmap_cubical_complex<T>(dimensions, top_dimensional_cells)
+            {
+                sizes_pdt.push_back(2*this->sizes[0]+1);
+                for(Simplex_handle i=1; i<this->dimension()-1; i++){           //In this loop we compute the product of the number of cubes in each direction, it optimizes the complexity of the get_coordinates_in_complex function
+                    sizes_pdt.push_back(sizes_pdt[i-1]*(2*this->sizes[i]+1));
+                }
+
+                std::cout << "Init embedding ...\n";
+                initalize_embedding();
+                initalize_embedding_index();
+                //print_embedding();
+                //print_filtration();
+                
+                std::cout << "Computing critical vertices\n";
+                if(num_jobs > std::thread::hardware_concurrency() || num_jobs <= 0){
+                    num_jobs = std::thread::hardware_concurrency();
+                }
+                compute_critical_vertices(num_jobs);
+                //print_critical_vertices_old();
+                //print_critical_vertices();
+                //print_critical_multiplicity();
```

Comment:
If these lines were used for a debugging/coding purpose, they should be removed in the final code.